File: ios/LookSeeProto/Models/HardNegativeUploadService.swift
//
//  HardNegativeUploadService.swift
//  LookSeeProto
//
//  Created by Ian Thompson on 6/16/26.
//

import Foundation
import Combine

@MainActor
final class HardNegativeUploadService: ObservableObject {
    @Published private(set) var status: String = String(localized: "Idle")
    @Published private(set) var progress: Double = 0
    @Published private(set) var isUploading = false

    private let baseURL = URL(string: "https://7gmn5z3uf2.execute-api.us-east-1.amazonaws.com/dev")!

    enum UploadError: LocalizedError {
        case noVideo
        case invalidURL
        case invalidResponse
        case badStatus(Int, String)
        case responseCountMismatch
        case missingLocalFile(String)
        case incompleteUpload

        var errorDescription: String? {
            switch self {
            case .noVideo: return String(localized: "No negative video was provided.")
            case .invalidURL: return String(localized: "The upload service returned an invalid URL.")
            case .invalidResponse: return String(localized: "The server returned an invalid response.")
            case .badStatus(let code, let body): return String(localized: "HTTP \(code): \(body)")
            case .responseCountMismatch: return String(localized: "Expected 1 upload URL, but received a different amount.")
            case .missingLocalFile(let filename): return String(localized: "The local video could not be found: \(filename)")
            case .incompleteUpload: return String(localized: "The upload failed to complete successfully.")
            }
        }
    }

    // MARK: - Public upload flow

    func upload(landmarkId: String, idToken: String, video: CapturedNegativeVideo) async throws -> HardNegativeCompleteResponse {
        
        isUploading = true
        progress = 0
        status = String(localized: "Preparing negative video…")

        defer { isUploading = false }

        do {
            let initResponse = try await initializeUpload(landmarkId: landmarkId, token: idToken, video: video)
            guard let uploadTarget = initResponse.uploads.first, initResponse.uploads.count == 1 else {
                throw UploadError.responseCountMismatch
            }

            progress = 0.1
            status = String(localized: "Uploading negative video to S3…")

            try await uploadVideo(video, to: uploadTarget)

            progress = 0.85
            status = String(localized: "Finalizing negative video…")

            let completeResponse = try await completeUpload(landmarkId: landmarkId, batchId: initResponse.batchId, negativeIds: [uploadTarget.negativeId], token: idToken)

            guard completeResponse.failedCount == 0, completeResponse.processedCount == 1 else {
                throw UploadError.incompleteUpload
            }

            progress = 1
            status = String(localized: "Negative video uploaded ✅")

            return completeResponse

        } catch {
            status = String(localized: "Negative upload failed: \(error.localizedDescription)")
            print("❌ Hard-negative upload failed:", error)
            throw error
        }
    }

    func reset() {
        status = String(localized: "Idle")
        progress = 0
        isUploading = false
    }

    // MARK: - Init

    private func initializeUpload(landmarkId: String, token: String, video: CapturedNegativeVideo) async throws -> HardNegativeInitResponse {
        let url = baseURL.appendingPathComponent("landmarks").appendingPathComponent(landmarkId).appendingPathComponent("hard-negatives").appendingPathComponent("init")
        let body = HardNegativeInitRequest(files: [HardNegativeFileRequest(filename: video.filename, contentType: "video/quicktime")])

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(token, forHTTPHeaderField: "Authorization")
        request.httpBody = try JSONEncoder().encode(body)

        let (data, response) = try await URLSession.shared.data(for: request)
        try validateAPIResponse(response, data: data)

        return try JSONDecoder().decode(HardNegativeInitResponse.self, from: data)
    }

    // MARK: - Presigned S3 uploads

    private func uploadVideo(_ video: CapturedNegativeVideo, to target: HardNegativeUploadTarget) async throws {
        guard FileManager.default.fileExists(atPath: video.fileURL.path) else {
            throw UploadError.missingLocalFile(video.filename)
        }
        guard let url = URL(string: target.uploadUrl) else {
            throw UploadError.invalidURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = "PUT"
        request.setValue(target.contentType, forHTTPHeaderField: "Content-Type")

        let (_, response) = try await URLSession.shared.upload(for: request, fromFile: video.fileURL)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw UploadError.invalidResponse
        }

        guard (200...299).contains(httpResponse.statusCode) else {
            throw UploadError.badStatus(httpResponse.statusCode, "S3 PUT failed for \(video.filename)")
        }
    }

    // MARK: - Complete

    private func completeUpload(landmarkId: String, batchId: String, negativeIds: [String], token: String) async throws -> HardNegativeCompleteResponse {
        let url = baseURL.appendingPathComponent("landmarks").appendingPathComponent(landmarkId).appendingPathComponent("hard-negatives").appendingPathComponent("complete")
        let body = HardNegativeCompleteRequest(batchId: batchId, negativeIds: negativeIds)

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(token, forHTTPHeaderField: "Authorization")
        request.httpBody = try JSONEncoder().encode(body)

        let (data, response) = try await URLSession.shared.data(for: request)
        try validateAPIResponse(response, data: data)

        return try JSONDecoder().decode(HardNegativeCompleteResponse.self, from: data)
    }

    private func validateAPIResponse(_ response: URLResponse, data: Data) throws {
        guard let httpResponse = response as? HTTPURLResponse else { throw UploadError.invalidResponse }
        guard (200...299).contains(httpResponse.statusCode) else {
            let body = String(data: data, encoding: .utf8) ?? ""
            throw UploadError.badStatus(httpResponse.statusCode, body)
        }
    }
}

// MARK: - Init models

private struct HardNegativeInitRequest: Encodable {
    let files: [HardNegativeFileRequest]
}

private struct HardNegativeFileRequest: Encodable {
    let filename: String
    let contentType: String
}

private struct HardNegativeInitResponse: Decodable {
    let message: String
    let batchId: String
    let landmarkId: String
    let landmarkLabel: String
    let landmarkFolder: String
    let expiresInSeconds: Int
    let uploads: [HardNegativeUploadTarget]
}

private struct HardNegativeUploadTarget: Decodable {
    let negativeId: String
    let uploadUrl: String
    let sourceBucket: String
    let sourceKey: String
    let contentType: String
}

// MARK: - Complete models

private struct HardNegativeCompleteRequest: Encodable {
    let batchId: String
    let negativeIds: [String]
}

struct HardNegativeCompleteResponse: Decodable {
    let message: String
    let landmarkId: String
    let batchId: String
    let processedCount: Int
    let failedCount: Int
    let dirtyMarked: Bool
    let processed: [HardNegativeProcessedItem]
    let failed: [HardNegativeFailedItem]
}

struct HardNegativeProcessedItem: Decodable {
    let negativeId: String
    let status: String
    let datasetImageKey: String
    let datasetLabelKey: String
}

struct HardNegativeFailedItem: Decodable {
    let negativeId: String
    let reason: String
}

File: Android/LookSee/app/src/main/java/looksee/angelll/com/models/HardNegativeUploadService.kt
package looksee.angelll.com.models

import com.google.gson.Gson
import java.io.File
import java.net.URLEncoder
import java.net.URL
import java.nio.charset.StandardCharsets
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

sealed class HardNegativeUploadError(message: String) : Exception(message) {
    data object NoVideo : HardNegativeUploadError("No negative video was provided.")
    data object InvalidUrl :
        HardNegativeUploadError("The upload service returned an invalid URL.")
    data object InvalidResponse :
        HardNegativeUploadError("The server returned an invalid response.")
    data class BadStatus(val code: Int, val responseBody: String) :
        HardNegativeUploadError("HTTP $code: $responseBody")
    data object ResponseCountMismatch :
        HardNegativeUploadError("Expected 1 upload URL, but received a different amount.")
    data class MissingLocalFile(val filename: String) :
        HardNegativeUploadError("The local video could not be found: $filename")
    data object IncompleteUpload :
        HardNegativeUploadError("The upload failed to complete successfully.")
}

class HardNegativeUploadService internal constructor(
    private val httpClient: UploadHttpClient,
    private val gson: Gson,
) {
    constructor() : this(UrlConnectionUploadHttpClient(), Gson())

    internal constructor(httpClient: UploadHttpClient) : this(httpClient, Gson())

    private val _status = MutableStateFlow("Idle")
    val status: StateFlow<String> = _status.asStateFlow()

    private val _progress = MutableStateFlow(0.0)
    val progress: StateFlow<Double> = _progress.asStateFlow()

    private val _isUploading = MutableStateFlow(false)
    val isUploading: StateFlow<Boolean> = _isUploading.asStateFlow()

    suspend fun upload(
        landmarkId: String,
        idToken: String,
        video: CapturedNegativeVideo,
    ): HardNegativeCompleteResponse {
        _isUploading.value = true
        _progress.value = 0.0
        _status.value = "Preparing negative video…"

        try {
            val initResponse = initializeUpload(landmarkId, idToken, video)
            if (initResponse.uploads.size != 1) {
                throw HardNegativeUploadError.ResponseCountMismatch
            }
            val uploadTarget = initResponse.uploads.single()

            _progress.value = 0.1
            _status.value = "Uploading negative video to S3…"
            uploadVideo(video, uploadTarget)

            _progress.value = 0.85
            _status.value = "Finalizing negative video…"
            val completeResponse = completeUpload(
                landmarkId = landmarkId,
                batchId = initResponse.batchId,
                negativeIds = listOf(uploadTarget.negativeId),
                token = idToken,
            )

            if (completeResponse.failedCount != 0 || completeResponse.processedCount != 1) {
                throw HardNegativeUploadError.IncompleteUpload
            }

            _progress.value = 1.0
            _status.value = "Negative video uploaded ✅"
            return completeResponse
        } catch (error: Throwable) {
            _status.value = "Negative upload failed: ${error.message.orEmpty()}"
            throw error
        } finally {
            _isUploading.value = false
        }
    }

    fun reset() {
        _status.value = "Idle"
        _progress.value = 0.0
        _isUploading.value = false
    }

    private suspend fun initializeUpload(
        landmarkId: String,
        token: String,
        video: CapturedNegativeVideo,
    ): HardNegativeInitResponse {
        if (!video.file.isFile) {
            throw HardNegativeUploadError.MissingLocalFile(video.filename)
        }
        val contentType = videoContentType(video.file)
        val body = HardNegativeInitRequest(
            files = listOf(HardNegativeFileRequest(video.filename, contentType)),
        )
        val response = httpClient.postJson(
            url = hardNegativeUrl(landmarkId, "init"),
            authorization = token,
            jsonBody = gson.toJson(body),
            timeoutMillis = API_TIMEOUT_MILLIS,
        )
        validateApiResponse(response)
        return decode(response.body, HardNegativeInitResponse::class.java)
    }

    private suspend fun uploadVideo(
        video: CapturedNegativeVideo,
        target: HardNegativeUploadTarget,
    ) {
        if (!video.file.isFile) {
            throw HardNegativeUploadError.MissingLocalFile(video.filename)
        }
        try {
            URL(target.uploadUrl)
        } catch (_: Exception) {
            throw HardNegativeUploadError.InvalidUrl
        }
        val response = httpClient.putFile(
            url = target.uploadUrl,
            contentType = target.contentType,
            file = video.file,
            timeoutMillis = MEDIA_UPLOAD_TIMEOUT_MILLIS,
        )
        if (response.statusCode !in 200..299) {
            throw HardNegativeUploadError.BadStatus(
                response.statusCode,
                "S3 PUT failed for ${video.filename}",
            )
        }
    }

    private suspend fun completeUpload(
        landmarkId: String,
        batchId: String,
        negativeIds: List<String>,
        token: String,
    ): HardNegativeCompleteResponse {
        val response = httpClient.postJson(
            url = hardNegativeUrl(landmarkId, "complete"),
            authorization = token,
            jsonBody = gson.toJson(HardNegativeCompleteRequest(batchId, negativeIds)),
            timeoutMillis = API_TIMEOUT_MILLIS,
        )
        validateApiResponse(response)
        return decode(response.body, HardNegativeCompleteResponse::class.java)
    }

    private fun hardNegativeUrl(landmarkId: String, action: String): String {
        val encodedId = URLEncoder.encode(
            landmarkId,
            StandardCharsets.UTF_8.name(),
        ).replace("+", "%20")
        return "$BASE_URL/landmarks/$encodedId/hard-negatives/$action"
    }

    private fun validateApiResponse(response: UploadHttpResponse) {
        if (response.statusCode !in 200..299) {
            throw HardNegativeUploadError.BadStatus(response.statusCode, response.body)
        }
    }

    private fun <T> decode(json: String, type: Class<T>): T = try {
        gson.fromJson(json, type) ?: throw HardNegativeUploadError.InvalidResponse
    } catch (error: HardNegativeUploadError) {
        throw error
    } catch (_: Exception) {
        throw HardNegativeUploadError.InvalidResponse
    }

    private fun videoContentType(file: File): String =
        if (file.extension.equals("mov", ignoreCase = true)) {
            "video/quicktime"
        } else {
            "video/mp4"
        }

    companion object {
        private const val BASE_URL =
            "https://7gmn5z3uf2.execute-api.us-east-1.amazonaws.com/dev"
        private const val API_TIMEOUT_MILLIS = 60_000
        private const val MEDIA_UPLOAD_TIMEOUT_MILLIS = 300_000
    }
}

private data class HardNegativeInitRequest(
    val files: List<HardNegativeFileRequest>,
)

private data class HardNegativeFileRequest(
    val filename: String,
    val contentType: String,
)

private data class HardNegativeInitResponse(
    val message: String,
    val batchId: String,
    val landmarkId: String,
    val landmarkLabel: String,
    val landmarkFolder: String,
    val expiresInSeconds: Int,
    val uploads: List<HardNegativeUploadTarget>,
)

private data class HardNegativeUploadTarget(
    val negativeId: String,
    val uploadUrl: String,
    val sourceBucket: String,
    val sourceKey: String,
    val contentType: String,
)

private data class HardNegativeCompleteRequest(
    val batchId: String,
    val negativeIds: List<String>,
)

data class HardNegativeCompleteResponse(
    val message: String,
    val landmarkId: String,
    val batchId: String,
    val processedCount: Int,
    val failedCount: Int,
    val dirtyMarked: Boolean,
    val processed: List<HardNegativeProcessedItem>,
    val failed: List<HardNegativeFailedItem>,
)

data class HardNegativeProcessedItem(
    val negativeId: String,
    val status: String,
    val datasetImageKey: String,
    val datasetLabelKey: String,
)

data class HardNegativeFailedItem(
    val negativeId: String,
    val reason: String,
)
